src/core/positioning/egenioussreceiver.cpp

```diff
@@ -0,0 +1,124 @@
+}
+
+void EgenioussReceiver::handleConnectDevice()
+{
+  mTcpSocket->connectToHost( mAddress, mPort, QTcpSocket::ReadWrite );
+}
+
+void EgenioussReceiver::connected()
+{
+  mSocketState = QAbstractSocket::ConnectedState;
+  emit socketStateChanged( mSocketState );
+  setValid( true );
+}
+
+void EgenioussReceiver::handleDisconnectDevice()
+{
+  mTcpSocket->disconnectFromHost();
+}
+
+void EgenioussReceiver::disconnected()
+{
+  if ( mTcpSocket->state() == QAbstractSocket::ConnectedState )
+  {
+    mSocketState = QAbstractSocket::UnconnectedState;
+    emit socketStateChanged( mSocketState );
+  }
+}
+
+QList<QPair<QString, QVariant>> EgenioussReceiver::details()
+{
+  QList<QPair<QString, QVariant>> detailsList;
+
+  if ( mPayload.isEmpty() )
+  {
+    return detailsList;
+  }
+
+  detailsList.append( qMakePair( "q", mPayload.value( "q" ).toDouble() ) );
+
+  return detailsList;
+}
+
+void EgenioussReceiver::onReadyRead()
+{
+  QByteArray mReceivedData = mTcpSocket->readAll();
+  if ( mReceivedData.size() < 9 )
+  {
+    return; // Received data is too short to process.
+  }
+  if ( static_cast<uint8_t>( mReceivedData[0] ) != 0xFE )
+  {
+    return; // Invalid start byte
+  }
+  uint32_t payloadLength;
+  QDataStream dataStream( mReceivedData.mid( 4, 4 ) );
+  dataStream.setByteOrder( QDataStream::LittleEndian );
+  dataStream >> payloadLength;
+  if ( mReceivedData.size() < 8 + payloadLength )
+  {
+    return; // Received data is too short to contain the payload.
+  }
+  QJsonDocument jsonDoc = QJsonDocument::fromJson( mReceivedData.mid( 8, payloadLength ) );
+  if ( jsonDoc.isNull() || !jsonDoc.isObject() )
+  {
+    return; // Failed to parse JSON
+  }
+  mPayload = jsonDoc.object();
+  const double latitude = mPayload.value( "lat" ).toDouble() == 0 ? std::numeric_limits<double>::quiet_NaN() : mPayload.value( "lat" ).toDouble();
+  const double longitude = mPayload.value( "lon" ).toDouble() == 0 ? std::numeric_limits<double>::quiet_NaN() : mPayload.value( "lon" ).toDouble();
+  const double elevation = mPayload.value( "alt" ).toDouble() == 0 ? std::numeric_limits<double>::quiet_NaN() : mPayload.value( "alt" ).toDouble();
+  mLastGnssPositionInformation = GnssPositionInformation(
+    latitude,
+    longitude,
+    elevation,
+    std::numeric_limits<double>::quiet_NaN(),
+    std::numeric_limits<double>::quiet_NaN(),
+    QList<QgsSatelliteInfo>(),
+    0,
+    0,
+    0,
+    std::numeric_limits<double>::quiet_NaN(),
+    std::numeric_limits<double>::quiet_NaN(),
+    QDateTime::fromMSecsSinceEpoch( mPayload.value( "utc" ).toDouble() / 1e6, Qt::UTC ),
+    QChar(),
+    0,
+    1 );
+
+  emit lastGnssPositionInformationChanged( mLastGnssPositionInformation );
+}
+
+void EgenioussReceiver::handleError( QAbstractSocket::SocketError error )
+{
+  switch ( error )
+  {
+    case QAbstractSocket::HostNotFoundError:
+      mLastError = tr( "Could not find the remote host" );
```

Comment:
I guess with the current interface here you should set mSocketState and mSocketStateString as well

Reply:
Correct.